Fix a regression in `group.present` in Windows

twangboy opened the pull request:
### What does this PR do?
Fixes a regression in `group.present` in Windows.
Caused by https://github.com/saltstack/salt/pull/43097

### What issues does this PR fix or reference?
Found in testing

### Previous Behavior
Because the state module is calling the execution module passing `gid` positionally it would fail with the following stack trace:
```
An exception occurred in this state: Traceback (most recent call last):
                File "C:\salt\bin\lib\site-packages\salt\state.py", line 1843, in call
                  **cdata['kwargs'])
                File "C:\salt\bin\lib\site-packages\salt\loader.py", line 1795, in wrapper
                  return f(*args, **kwargs)
                File "C:\salt\bin\lib\site-packages\salt\states\group.py", line 249, in present
                  system=system):
              TypeError: add() takes exactly 1 argument (3 given)
```

### New Behavior
State module passes `gid` as a keyword arg.

### Tests written?
No

twangboy committed:
Accept *args

twangboy committed:
Remove *args, pass gid as a keyword

twangboy committed:
Account for 15 character limit in hostname

twangboy committed:
Document 15 character limit

rallytime commented on salt/utils/win_functions.py:
```diff
@@ -152,8 +152,10 @@ def get_sam_name(username):
     '''
     if '\\' not in username:
-        username = '{0}\\{1}'.format(platform.node(), username)
+        username = '{0}\\{1}'.format(platform.node()[:15], username)
```
Is this change related to the `gid=gid` fix? I don't see an explanation of why this part is included in this PR or why this would be necessary.

If it should be included here, can you add a some documentation or an explanation of why this is truncated? That will help the will code maintenance in the future.

cachedout replied:
Can you also explain why truncated this is needed? This seems like it would really potentially cause a lot of confusion and potential duplication for users of this function.

twangboy replied:
@rallytime Yeah, this is also a part of the fix. When I originally "Fixed" this module, I was on a system with a shorter hostname, so I didn't see the issue with longer hostnames.

@cachedout The problem is that in the execution module when it lists the members of a group it lists them as `hostname\\groupname`. The hostname is truncated to 15 characters. So when the state tries to figure out if the user is already in the group it thinks it's not because it's looking for the `fullhostname\\username` instead of the truncated one. The `win_function.get_sam_name` was written specifically for the group thing on Windows.

damon-atkins replied:
Windows does not permit computer names that exceed 15 characters.  The maximum length of the host name and of the fully qualified domain name (FQDN) is 63 bytes per label and 255 characters per FQDN. The latter is based on the maximum path length possible with an Active Directory Domain name with the paths needed in SYSVOL, and this needs to obey to the 260 character MAX_PATH limitation.  https://support.microsoft.com/en-au/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and

Not sure why name would be allowed, unless then name being returned has a domain in it. (which maybe it needs to be stripped off)  You can have long usernames but they must be in the format of <username>@<realm>  where realm is <DOMAIN>.<rest of DNS Domain>

damon-atkins replied:
win32security.TranslateName might be usefull

twangboy replied:
`win32security.TranslateName` seems to only work when the host is joined to a domain

damon-atkins replied:
`sid = win32security.GetBinarySid(uid)` # This is not correct as it excepts  SID not a UserName but may give you a glue.
`name, domain, account_type = win32security.LookupAccountSid(None, sid)`

damon-atkins replied:
Backwards and then Forward lookup
```
sid_str = win32security.LookupAccountName(None, username)[0]
sid = win32security.GetBinarySid(sid_str)
name, domain, account_type = win32security.LookupAccountSid(None, sid)
```
?

twangboy replied:
That might work...

cachedout commented on salt/states/group.py:
```diff
@@ -244,9 +244,7 @@ def present(name,
-        if __salt__['group.add'](name,
-                                 gid,
-                                 system=system):
+        if __salt__['group.add'](name, gid=gid, system=system):
```
👍

twangboy committed:
Fix example in function docs for get_sam_name

twangboy committed:
Improve get_sam_name

Return the actual sam name as reported by LookupAccountName
If the user does not exist, return a fabricated SAM name
Move lower to the state module

twangboy commented:
Rebased...

rallytime commented:
@twangboy This has a pylint error in it. Can you fix that really quickly in a new PR?

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/15823/violations/file/salt/utils/win_functions.py/
